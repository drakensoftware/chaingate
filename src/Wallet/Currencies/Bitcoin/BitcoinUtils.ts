import Decimal from 'decimal.js'

export enum AddressType {
    PubKeyHash,
    ScriptHash ,
    WitnessPubKeyHash,
    WitnessScriptHash,
    TapRoot
}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace BitcoinUtils {

    export function toSatoshi(amount: Decimal): Decimal {
        return amount.mul(new Decimal(1_0000_0000))
    }
}
