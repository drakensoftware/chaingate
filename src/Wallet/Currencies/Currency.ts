import Decimal from 'decimal.js'
import {Api} from 'chaingate-client'
import {Phrase, PrivateKey, PrivateKeySource, PublicKey, Seed} from '../Keys'

export class NotEnoughFundsError extends Error {
    constructor(currency: Currency, missingFunds: CurrencyAmount) {
        super(`Wallet does not have enough funds. ${missingFunds.baseAmount} ${missingFunds.baseSymbol} needed.`)
        if (Error.captureStackTrace) Error.captureStackTrace(this, NotEnoughFundsError)
        this.name = this.constructor.name
    }
}

export class CannotDerive extends Error {
    constructor() {
        super('Derivation path are not supported (Are you using an imported private key?)')
        if (Error.captureStackTrace) Error.captureStackTrace(this, CannotDerive)
        this.name = this.constructor.name
    }
}

export class CannotParseAmount extends Error {
    constructor(amountStr: string) {
        super(`Invalid amount: ${amountStr}`)
        if (Error.captureStackTrace) Error.captureStackTrace(this, CannotDerive)
        this.name = this.constructor.name
    }
}

export type FeeGrade = 'low' | 'normal' | 'high' | 'maximum'

export abstract class ConfirmedTransaction {
    abstract readonly txId: string

    abstract isConfirmed(): Promise<boolean>
    waitToBeConfirmed(): Promise<void> {
        // eslint-disable-next-line no-async-promise-executor
        return new Promise(async (resolve) => {
            // First check before starting the interval
            if (await this.isConfirmed()) {
                resolve()
            } else {
                const intervalId = setInterval(async () => {
                    if (await this.isConfirmed()) {
                        clearInterval(intervalId)
                        resolve()
                    }
                }, 10000)
            }
        })
    }
}

export abstract class PreparedTransaction {
    abstract readonly possibleFees: { [key in FeeGrade]: Fee }
    abstract confirm(fee: string | Fee): Promise<ConfirmedTransaction>
}

export class CurrencyAmount {
    readonly baseAmount: Decimal
    readonly baseSymbol: string

    readonly minimalUnitAmount: Decimal
    readonly minimalUnitSymbol: string

    constructor(currency: Currency, baseAmount: Decimal){
        this.baseAmount = baseAmount
        this.minimalUnitAmount = baseAmount.mul( (new Decimal(10)).pow(currency.currencyInfo.decimals))

        this.baseSymbol = currency.currencyInfo.symbol
        this.minimalUnitSymbol = currency.currencyInfo.minimalUnitSymbol
    }
}

export abstract class Fee {}
export abstract class FeeInitializer {}
export type PossibleFees<T> = {[key in FeeGrade]: T}

export type CurrencyInfo = {
    id: string
    name: string
    defaultDerivationPath: string
    svgLogoUrl: string,
    symbol: string,
    minimalUnitSymbol: string,
    decimals: number
}

export abstract class Currency{
    protected readonly api: Api
    public readonly currencyInfo: CurrencyInfo

    protected publicKey: PublicKey | null
    protected privateKeySource: PrivateKeySource

    protected constructor(currencyInfo: CurrencyInfo, api: Api, privateKeySource: PrivateKeySource) {
        this.currencyInfo = currencyInfo
        this.api = api
        this.privateKeySource = privateKeySource
        this._derivationPath = currencyInfo.defaultDerivationPath
    }

    async getPrivateKey() {
        if (this.privateKeySource instanceof PrivateKey) return this.privateKeySource
        else if (this.privateKeySource instanceof Seed || this.privateKeySource instanceof Phrase) {
            return this.privateKeySource.derive(this.getDerivationPath())
        }
    }

    async initializePublicKey(){
        this.publicKey = await (await this.getPrivateKey()).getPublicKey()
    }

    amount(amountStr: string) : CurrencyAmount{
        const regex = /^(\d+(\.\d+)?) ?([A-Za-z]+)?$/
        const match = amountStr.match(regex)
        if (!match) throw new CannotParseAmount(amountStr)

        const amount = new Decimal(match[1])

        if (match[3]) {
            let currencySymbol = match[3].toLowerCase()
            if(currencySymbol.endsWith('s')) currencySymbol = currencySymbol.substring(currencySymbol.length - 1) //Remove plural -s

            if (currencySymbol === this.currencyInfo.symbol.toLowerCase() ||
                currencySymbol === this.currencyInfo.name.toLowerCase()) {
                return new CurrencyAmount(this, amount)
            } else if (currencySymbol === this.currencyInfo.minimalUnitSymbol.toLowerCase()) {
                return new CurrencyAmount(this, amount.mul((new Decimal(10)).pow(this.currencyInfo.decimals)))
            } else {
                throw new CannotParseAmount(amountStr)
            }
        }

        return new CurrencyAmount(this, amount)
    }


    abstract fee(fee: unknown): Fee

    abstract getAddress(): Promise<string>

    abstract getBalance(address?: string): Promise<{confirmed: CurrencyAmount, unconfirmed: CurrencyAmount}>

    abstract prepareTransfer(toAddress: string, amount: CurrencyAmount): Promise<PreparedTransaction>

    // #region Derivation Paths
    private _derivationPath: string | null

    getDerivationPath(){
        return this._derivationPath
    }

    async setDerivationPath(newDerivationPath: string){
        this._derivationPath = newDerivationPath
        await this.initializePublicKey()
    }
    // #endregion
}
