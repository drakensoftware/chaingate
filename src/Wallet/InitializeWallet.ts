import {PhaseNumOfWords, PhraseGenerator, PhraseLanguage} from '../Utils/PhraseGenerator'
import {PhraseWallet} from './PhraseWallet'
import {ChainGateClient} from 'chaingate-client'
import {Phrase} from './Keys'
import {decryptKeystoreV1, isKeystoreV1} from './Crypto/KeystoreDecoder/KeystoreV1'
import {decryptKeystore} from './Crypto/KeystoreDecoder/OtherKeystores'
import {ImportedPrivateKey} from './ImportedPrivateKey'
import {PrivateKey} from './Keys'
import IncorrectPassword from './Crypto/KeystoreDecoder/IncorrectPassword'
import {Utils} from '../Utils'
import {HDWallet} from './HDWallet'
import {Seed} from './Keys'
import wif from 'wif'

export class EncodingError extends Error {
    constructor(message: string) {
        super(message)
        if (Error.captureStackTrace) Error.captureStackTrace(this, EncodingError)
        this.name = this.constructor.name
    }
}

export async function create(apiKey: string,
    phraseLanguage: PhraseLanguage = 'english', phraseNumOfWords: PhaseNumOfWords = 12) {
    const phrase = PhraseGenerator.generateNewPhrase(phraseLanguage, phraseNumOfWords)
    return await PhraseWallet.new(new ChainGateClient(apiKey), new Phrase(phrase))
}

export async function fromKeystore(apiKey: string, keystore: string, password: string) {
    //Some keystores store the phrase and others the private key
    try {
        const decodedKeystore = isKeystoreV1(keystore)
            ? await decryptKeystoreV1(keystore, password)
            : await decryptKeystore(keystore, password)

        let phrase
        try {
            phrase = new TextDecoder().decode(decodedKeystore)
        } catch { /* empty */ }

        if (phrase && Phrase.isValidPhrase(phrase)) {
            return PhraseWallet.new(new ChainGateClient(apiKey), new Phrase(phrase))
        } else {
            return ImportedPrivateKey.new(new ChainGateClient(apiKey), new PrivateKey(decodedKeystore))
        }
    } catch (ex) {
        if (ex instanceof IncorrectPassword) throw ex
        throw new EncodingError('Invalid keystore')
    }
}

export async function fromPhrase(apiKey: string, phrase: string) {
    if(!Phrase.isValidPhrase(phrase)) throw new EncodingError('Invalid phrase')
    return PhraseWallet.new(new ChainGateClient(apiKey), new Phrase(phrase))
}

export async function fromSeed(apiKey: string, seed: string | Uint8Array, ){
    if(typeof seed === 'string' && Utils.isHex(seed)) seed = Utils.hexToBytes(seed)
    else if(typeof seed === 'string') throw new EncodingError('Invalid seed')

    return HDWallet.new(new ChainGateClient(apiKey), new Seed(seed))
}

export async function  fromPrivateKey(apiKey: string, privateKey: string | Uint8Array){
    if(typeof privateKey === 'string' && Utils.isHex(privateKey)) privateKey = Utils.hexToBytes(privateKey)
    else if(typeof privateKey === 'string' && Utils.isBase58(privateKey)){
        try{
            privateKey = new Uint8Array(wif.decode(privateKey).privateKey)
        }catch (_ex){
            throw new EncodingError('The string supplied in Wallet Import Format (WIF) is deemed to be invalid')
        }
    }else if(typeof privateKey === 'string') throw new EncodingError('Private key is invalid or the format is unrecognized')

    return ImportedPrivateKey.new(new ChainGateClient(apiKey), new PrivateKey(privateKey))
}
