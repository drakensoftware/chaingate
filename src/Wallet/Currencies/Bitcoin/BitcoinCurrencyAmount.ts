import {CurrencyAmount} from '../Currency'
import Decimal from 'decimal.js'

export default class BitcoinCurrencyAmount extends CurrencyAmount {
    private _baseAmount: Decimal

    get baseAmount(): Decimal {
        return this._baseAmount
    }
    get baseSymbol(): string {
        return 'BTC'
    }

    get minimalAmount(): Decimal {
        return this._baseAmount.mul(1_0000_0000)
    }

    get minimalSymbol(): string {
        return 'satoshis'
    }

    get btc(): Decimal {
        return this._baseAmount
    }

    get satoshis(): Decimal {
        return this._baseAmount.mul(1_0000_0000)
    }

    constructor(baseAmount: Decimal) {
        super()
        this._baseAmount = baseAmount
    }
}
