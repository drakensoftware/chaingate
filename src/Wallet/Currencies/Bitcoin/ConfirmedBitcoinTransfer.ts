import {ConfirmedTransaction} from '../Currency'
import {ChainGateClient} from 'chaingate-client'
import {ConsumeFunction} from '../../../CGDriver'

export class ConfirmedBitcoinTransaction extends ConfirmedTransaction{
    public readonly txId: string
    private readonly chainGateClient: ChainGateClient

    constructor(chainGateClient: ChainGateClient, txId: string) {
        super()
        this.chainGateClient = chainGateClient
        this.txId = txId
    }

    async isConfirmed(): Promise<boolean> {
        try{
            const transactionDetails = await ConsumeFunction(
                this.chainGateClient.Bitcoin,
                this.chainGateClient.Bitcoin.transactionDetails,
                this.txId)
            return transactionDetails.blockHeight != null
        }catch (_ex){
            return false
        }
    }
}
