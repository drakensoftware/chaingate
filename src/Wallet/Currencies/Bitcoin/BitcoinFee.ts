import BitcoinCurrencyAmount from './BitcoinCurrencyAmount'

export enum FeeKind {
    FeeAmount,
    FeePerByte
}

export class BitcoinFee{
    public readonly fee: BitcoinCurrencyAmount
    public readonly kind: FeeKind

    private constructor(fee: BitcoinCurrencyAmount, feeKind: FeeKind){
        this.fee = fee
        this.kind = feeKind
    }

    public static feePerByte(feePerByte: BitcoinCurrencyAmount): BitcoinFee{
        return new BitcoinFee(feePerByte, FeeKind.FeePerByte)
    }

    public static feeAmount(fee: BitcoinCurrencyAmount): BitcoinFee{
        return new BitcoinFee(fee, FeeKind.FeeAmount)
    }

}
