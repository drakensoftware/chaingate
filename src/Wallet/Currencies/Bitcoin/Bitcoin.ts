import Decimal from 'decimal.js'
import {Currency, NotEnoughFundsError, PossibleFees} from '../Currency'
import {ChainGateClient} from 'chaingate-client'
import {PrivateKeySource} from '../../Keys'
import {AddressType} from './BitcoinUtils'
import {Address as TapScriptAddress, Tap} from '@cmdcode/tapscript'
import {ConsumeFunction} from '../../../CGDriver'
import BitcoinCurrencyAmount from './BitcoinCurrencyAmount'
import {PreparedBitcoinTransfer} from './PreparedBitcoinTransfer'
import {Utxos} from 'chaingate-client/dist/api'
import {BitcoinFee} from './BitcoinFee'
import * as Utils from '../../../Utils/Utils'

export type UTXO = {
    amount: Decimal, txid: string, n: number, script: Uint8Array
}

export class Bitcoin extends Currency {
    constructor(chainGateClient: ChainGateClient, privateKeySource: PrivateKeySource) {
        super({
            symbol: 'BTC',
            id: 'bitcoin',
            defaultDerivationPath: 'm/84\'/0\'/0\'/0/0',
            name: 'bitcoin',
            svgLogoUrl: Utils.buildUrlWithApiKey('https://api.chaingate.dev/bitcoin/logo')
        }
        ,
        chainGateClient,
        privateKeySource)
    }

    getAddress() {
        return this.getAddressOfType(AddressType.WitnessPubKeyHash)
    }

    async getAddressOfType(typeOfAddress: AddressType) {
        switch (typeOfAddress) {
        case AddressType.PubKeyHash:
            return TapScriptAddress.p2pkh.fromPubKey(this.publicKey.compressed)
        case AddressType.WitnessPubKeyHash:
            return TapScriptAddress.p2wpkh.fromPubKey(this.publicKey.compressed)
        case AddressType.TapRoot:
            return TapScriptAddress.p2tr.fromPubKey(Tap.getPubKey(this.publicKey.compressed)[0])
        case AddressType.ScriptHash:
        case AddressType.WitnessScriptHash:
            throw new Error('Generating script addresses is currently unsupported')
        }
    }

    async getBalance(address?: string) {
        if (!address) address = await this.getAddress()
        const balance = await ConsumeFunction(
            this.chaingateClient.Bitcoin,
            this.chaingateClient.Bitcoin.addressBalance,
            address)
        return {
            confirmed: new BitcoinCurrencyAmount(new Decimal(balance.confirmed)),
            unconfirmed: new BitcoinCurrencyAmount(new Decimal(balance.unconfirmed))
        }
    }

    async prepareTransfer(toAddress: string, amount: BitcoinCurrencyAmount): Promise<PreparedBitcoinTransfer> {
        if (toAddress.startsWith('bcrt1q')) throw new Error('Taproot addresses are not yet supported')

        const fromAddress = await this.getAddress()
        const utxos = await this.buildUtxos(amount.btc)
        const possibleFees = await this.buildPossibleFees()

        return new PreparedBitcoinTransfer(this.chaingateClient, this, fromAddress, toAddress, amount, utxos, possibleFees)
    }

    private async buildUtxos(amountBtc: Decimal) {
        const pickedUtxos = []
        let totalUtxosAmount = new Decimal(0)

        let page = 0
        let utxosByAddress: Utxos
        do {
            utxosByAddress = await ConsumeFunction(
                this.chaingateClient.Bitcoin,
                this.chaingateClient.Bitcoin.utxosByAddress,
                await this.getAddress(), page)

            for (const utxo of utxosByAddress.utxos) {
                pickedUtxos.push({
                    txid: utxo.hash, amount: new Decimal(utxo.amount), n: utxo.n,
                    script: Utils.hexToBytes(utxo.script)
                })
                totalUtxosAmount = totalUtxosAmount.add(new Decimal(utxo.amount))

                if (totalUtxosAmount.gte(amountBtc)) break
            }

            page++
        } while (totalUtxosAmount.lt(amountBtc) && utxosByAddress.page != utxosByAddress.lastPage)

        if (totalUtxosAmount.lt(amountBtc)) throw new NotEnoughFundsError(this, new BitcoinCurrencyAmount(amountBtc.minus(totalUtxosAmount)))

        return pickedUtxos
    }

    private async buildPossibleFees(): Promise<PossibleFees<BitcoinFee>> {
        const feeRateResponse = await ConsumeFunction(
            this.chaingateClient.Bitcoin,
            this.chaingateClient.Bitcoin.feeRate)

        return {
            low: BitcoinFee.feePerByte(new BitcoinCurrencyAmount(new Decimal(feeRateResponse['low'].feePerByte))),
            normal: BitcoinFee.feePerByte(new BitcoinCurrencyAmount(new Decimal(feeRateResponse['normal'].feePerByte))),
            high: BitcoinFee.feePerByte(new BitcoinCurrencyAmount(new Decimal(feeRateResponse['high'].feePerByte))),
            maximum: BitcoinFee.feePerByte(new BitcoinCurrencyAmount(new Decimal(feeRateResponse['maximum'].feePerByte))),
        }
    }
}
