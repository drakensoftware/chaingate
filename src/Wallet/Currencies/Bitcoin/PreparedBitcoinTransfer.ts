import {FeeGrade, PossibleFees, PreparedTransaction} from '../Currency'
import {Transaction} from 'bitcore-lib'
import {ConfirmedBitcoinTransaction} from './ConfirmedBitcoinTransfer'
import {Utils} from '../../../Utils'
import {ChainGateClient} from 'chaingate-client'
import {BitcoinFee, FeeKind} from './BitcoinFee'
import {BitcoinUtils} from './BitcoinUtils'
import BitcoinCurrencyAmount from './BitcoinCurrencyAmount'
import {Bitcoin} from './index'
import {ConsumeFunction} from '../../../CGDriver'
import {UTXO} from './Bitcoin'
import UnspentOutput = Transaction.UnspentOutput;

export class PreparedBitcoinTransfer extends PreparedTransaction {
    private readonly chaingateClient: ChainGateClient
    private readonly bitcoin: Bitcoin
    private fromAddress: string
    private toAddress: string
    private amount: BitcoinCurrencyAmount
    private utxos: Array<UTXO>
    readonly possibleFees: PossibleFees<BitcoinFee>

    constructor(chainGateClient: ChainGateClient,
        bitcoin: Bitcoin,
        fromAddress: string,
        toAddress: string,
        amount: BitcoinCurrencyAmount,
        utxos: Array<UTXO>,
        possibleFees: PossibleFees<BitcoinFee>
    ) {
        super()
        this.chaingateClient = chainGateClient
        this.bitcoin = bitcoin
        this.fromAddress = fromAddress
        this.toAddress = toAddress
        this.amount = amount
        this.utxos = utxos
        this.possibleFees = possibleFees
    }

    async confirm(fee: FeeGrade | BitcoinFee): Promise<ConfirmedBitcoinTransaction> {
        if (!(fee instanceof BitcoinFee)) fee = this.possibleFees[fee]

        const bitcoreUtxo = this.utxos.map(t => new UnspentOutput({
            address: this.fromAddress,
            txId: t.txid,
            outputIndex: t.n,
            script: Utils.bytesToHex(t.script, false),
            satoshis: BitcoinUtils.toSatoshi(t.amount).toNumber(),
        }))

        let transaction = new Transaction()
            .from(bitcoreUtxo)
            .to(this.toAddress, this.amount.satoshis.toNumber())
            .change(this.fromAddress)

        if(fee.kind == FeeKind.FeeAmount)
            transaction = transaction.fee(fee.fee.satoshis.toNumber())
        if(fee.kind == FeeKind.FeePerByte)
            transaction = transaction.feePerKb(fee.fee.satoshis.mul(1_024).toNumber())

        transaction = transaction.sign(Utils.bytesToHex(((await this.bitcoin.getPrivateKey()).raw), false))

        const broadcastTx = await ConsumeFunction(
            this.chaingateClient.Bitcoin,
            this.chaingateClient.Bitcoin.broadcastTransaction,
            {transactionRaw: transaction.serialize()}
        )

        return new ConfirmedBitcoinTransaction(this.chaingateClient, broadcastTx.txId)
    }

}
