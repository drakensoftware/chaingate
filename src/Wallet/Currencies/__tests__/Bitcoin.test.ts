import {AddressType} from '../Bitcoin/BitcoinUtils'
import {NotEnoughFundsError} from '../Currency'
import {Btc} from '../../Units'
import * as InitializeWallet from '../../InitializeWallet'

describe('Bitcoin', function () {
    it('Generate addresses of different types', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)

        const mnemonic = 'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about'
        const wallet = await InitializeWallet.fromPhrase('', mnemonic)

        expect(await wallet.currencies.bitcoin.getAddress())
            .toBe('bc1qcr8te4kr609gcawutmrza0j4xv80jy8z306fyu')

        expect(await wallet.currencies.bitcoin.getAddressOfType(AddressType.WitnessPubKeyHash))
            .toBe('bc1qcr8te4kr609gcawutmrza0j4xv80jy8z306fyu')

        await wallet.currencies.bitcoin.setDerivationPath('m/44\'/0\'/0\'/0/0')
        expect(await wallet.currencies.bitcoin.getAddressOfType(AddressType.PubKeyHash))
            .toBe('1LqBGSKuX5yYUonjxT5qGfpUsXKYYWeabA')

        await wallet.currencies.bitcoin.setDerivationPath('m/86\'/0\'/0\'/0/0')
        expect(await wallet.currencies.bitcoin.getAddressOfType(AddressType.TapRoot))
            .toBe('bc1p5cyxnuxmeuwuvkwfem96lqzszd02n6xdcjrs20cac6yqjjwudpxqkedrcr')

        await expect(wallet.currencies.bitcoin.getAddressOfType(AddressType.ScriptHash)).rejects.toThrow()
        await expect(wallet.currencies.bitcoin.getAddressOfType(AddressType.WitnessScriptHash)).rejects.toThrow()
    })

    it('Get address balance', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)

        const mnemonic = 'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about'
        const wallet = await InitializeWallet.fromPhrase('', mnemonic)

        expect(await wallet.currencies.bitcoin.getAddress()).toBe('bc1qcr8te4kr609gcawutmrza0j4xv80jy8z306fyu')

        const balance = await wallet.currencies.bitcoin.getBalance()
        expect(balance.confirmed.btc.toNumber()).toBe(0)
        expect(balance.unconfirmed.btc.toNumber()).toBe(0)
    })

    it('Transfer some bitcoins', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)

        const mnemonic = 'zero sun street chronic online omit buzz critic south sample video drastic base provide cream'
        const wallet = await InitializeWallet.fromPhrase('', mnemonic)

        expect(await wallet.currencies.bitcoin.getAddress()).toBe('bc1ql0gj7t80sxgg5x9s5ft4wsdehhd8rhrpmtfg09')

        const preparedTransfer = await wallet.currencies.bitcoin.prepareTransfer('bc1qv0tvnnvssfqaglmms6tmpwnrl3rq2mrjvq2z7k', Btc('0.00039'))
        expect(preparedTransfer.possibleFees).toMatchSnapshot()
        const result = await preparedTransfer.confirm('normal')
        expect(result.txId).toBe('619e19924cce832c08e648624b4bcd3e5c21f6daf9c24c65042b4904083cb645')
        await result.waitToBeConfirmed()
        expect(await result.isConfirmed()).toBeTruthy()
    })

    it('Transfer some bitcoins (not enough funds)', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)

        const mnemonic = 'zero sun street chronic online omit buzz critic south sample video drastic base provide cream'
        const wallet = await InitializeWallet.fromPhrase('', mnemonic)

        expect(await wallet.currencies.bitcoin.getAddress()).toBe('bc1ql0gj7t80sxgg5x9s5ft4wsdehhd8rhrpmtfg09')

        await expect(async () =>
            await wallet.currencies.bitcoin.prepareTransfer('bc1qv0tvnnvssfqaglmms6tmpwnrl3rq2mrjvq2z7k', Btc('1'))).rejects
            .toThrow(NotEnoughFundsError)
    })

})
