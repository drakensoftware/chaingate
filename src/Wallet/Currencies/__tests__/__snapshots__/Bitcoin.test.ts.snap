// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bitcoin Transfer some bitcoins 1`] = `
{
  "high": BitcoinFee {
    "fee": BitcoinCurrencyAmount {
      "_baseAmount": "0.00000122406",
    },
    "kind": 1,
  },
  "low": BitcoinFee {
    "fee": BitcoinCurrencyAmount {
      "_baseAmount": "0.00000105456",
    },
    "kind": 1,
  },
  "maximum": BitcoinFee {
    "fee": BitcoinCurrencyAmount {
      "_baseAmount": "0.00000122406",
    },
    "kind": 1,
  },
  "normal": BitcoinFee {
    "fee": BitcoinCurrencyAmount {
      "_baseAmount": "0.00000122406",
    },
    "kind": 1,
  },
}
`;
